### PR TITLE
version number cleanup

### PR DESCRIPTION
**Pull request by LukeShu**
## Description

Ever since 5d5f6e3d8de2d4c950615ed5f96ba11811f932c4, `-rcN` tags don't actually get pushed to quay.io.

I could just revert that commit, and then say `${GIT_TAG_SANITIZED#v}`, but IMO the root problem is that there are a whole buncha version-related variables, and no one really knows what they're each for, at least not for more than 10 minutes at a time.

So, instead of doing the bare minimum to fix the docker-tag issue, go ahead and try to clean up the version-variable situation.

Oh, while we're at it, use `docker.mk` to do the RC and EA docker pushes, since that makes things easier (it's almost like I had ambassador.git's use-case in mind when I wrote the current version of `docker.mk`).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -26,58 +26,48 @@ SHELL = bash
...
 _git_remote_urls := $(shell git remote | xargs -n1 git remote get-url --all)
 IS_PRIVATE ?= $(findstring private,$(_git_remote_urls))
 
-# Note that for everything except RC builds, VERSION will be set to the version
-# we'd use for a GA build. This is by design.
+# RELEASE_VERSION is an X.Y.Z (semver) string is what we will
+# upload/release the image as.  It does NOT include a leading 'v'
+# (trimming the 'v' from the git tag is what the 'patsubst' is for).
+# If this is an RC or EA, then it's includes the '-rcN' or '-eaN'
+# suffix.
```
Comment (kflynn) [inline Makefile]: This comment is wonderful except that there's a lot of abuse of 'it' and 'is' in it. 🙂

File: Makefile [hunk 2]
```diff
@@ -26,58 +26,48 @@ SHELL = bash
...
-# Also note that we strip off the leading 'v' here -- that's just for the git tag.
-ifneq ($(GIT_TAG_SANITIZED),)
-VERSION = $(patsubst v%,%,$(firstword $(subst -, ,$(GIT_TAG_SANITIZED))))
-else
-VERSION = $(patsubst v%,%,$(firstword $(subst -, ,$(GIT_VERSION))))
-endif
-
-# We need this for tagging in some situations.
-LATEST_RC=$(VERSION)-rc-latest
+# BUILD_VERSION is of the the same thing, but is the version number
```
Comment (kflynn) [inline Makefile]: `s/is of the/is the/` ??

File: Makefile [hunk 3]
```diff
@@ -724,15 +712,31 @@ release-prep:
...
+	docker pull $(AMBASSADOR_DOCKER_REPO):$(RELEASE_VERSION)-rc-latest
+	docker tag $(AMBASSADOR_DOCKER_REPO):$(RELEASE_VERSION)-rc-latest $(AMBASSADOR_DOCKER_REPO):$(RELEASE_VERSION)
+	docker push $(AMBASSADOR_DOCKER_REPO):$(RELEASE_VERSION)
 	$(MAKE) SCOUT_APP_KEY=app.json STABLE_TXT_KEY=stable.txt update-aws
 
+#release-ga: RETAG_INSTEAD_OF_REBUILD = true
+#release-ga: ambassadr.docker.push.release-ga
+#release-ga: SCOUT_APP_KEY = app.json
+#release-ga: STABLE_TXT_KEY = stable.txt
+#release-ga: update-aws
```
Comment (kflynn) [inline Makefile]: ??

File: Makefile [hunk 4]
```diff
@@ -724,15 +712,31 @@ release-prep:
...
+
+release-rc: ambassadr.docker.push.release-rc
```
Comment (kflynn) [inline Makefile]: In this, in `release-rc`, and in `release-ga`, are you intentionally dropping the 'o' in `ambassador` ?

File: releng/travis-script.sh [hunk 5]
```diff
@@ -19,32 +19,27 @@ set -o nounset
...
+    elif [[ "$TRAVIS_TAG" =~ ^v[0-9]+\.[0-9]+\.[0-9]+-rc[0-9]+$ ]]; then
+        COMMIT_TYPE=RC
+    elif [[ "$TRAVIS_TAG" =~ ^v[0-9]+\.[0-9]+\.[0-9]+-ea[0-9]+$ ]]; then
+        COMMIT_TYPE=EA
+    else
+        echo "TRAVIS_TAG '$TRAVIS_TAG' is not in one of the recognized tag formats:" >&2
+        echo " - 'vSEMVER'" >&2
+        echo " - 'vSEMVER-rcN'" >&2
+        echo " - 'vSEMVER-eaN'" >&2
+        echo "Note that the tag name must start with a lowercase 'v' (which the Makefile will trim off)" >&2
```
Comment (kflynn) [inline releng/travis-script.sh]: I think I'd drop "(which the Makefile will trim off)" here...

File: Makefile (hunk 2 above)
Comment (LukeShu) [reply Makefile]: I think I started to write "is of the same format".

**Comment by LukeShu** [comment]
Amended.  I guess that's what happens when I try to rush a PR before I head home.

**Comment by LukeShu** [comment]
```console
$ export DOCKER_REGISTRY=- DOCKER_EXTERNAL_REGISTRY=quay.io/datawire

$ git describe --tags
v0.82.0-61-gddc9f9e2

$ make ambassador.docker
…
Successfully tagged ambassador:0.82.0-61-gddc9f9e2

$ grep -A5 -B4 =GitInfo ambassador/ambassador/VERSION.py
Version = "0.82.0-61-gddc9f9e2"

Build = BuildInfo(
    version=Version,
    git=GitInfo(
        commit="ddc9f9e2",
        branch="lukeshu/versions",
        dirty=bool(""),
        description="v0.82.0-61-gddc9f9e2",
    )

$ git tag v0.82.1-ea1

$ make ambassador.docker
…
Successfully tagged ambassador:0.82.1-ea1

$ make ambassador.docker.tag.release-ea
…
  * 
Successfully tagged ambassador:0.82.1-ea1
printf '%s\n' $(cat ambassador.docker) quay.io/datawire/ambassador:0.82.1-ea1 | /home/lukeshu/src/github.com/datawire/ambassador/build-aux/bin/write-dockertagfile ambassador.docker.tag.release-ea

$ cat ambassador.docker.tag.release-ea
sha256:3935dd26da396bcb186db8fa5a0a1afed5d2fe23f1d3ab681898ba0c426bf555
quay.io/datawire/ambassador:0.82.1-ea1

$ git tag -d v0.82.1-ea1

$ git tag v0.82.1-rc1

$ make ambassador.docker.tag.release-rc
…
Successfully tagged ambassador:0.82.1-rc1
printf '%s\n' $(cat ambassador.docker) quay.io/datawire/ambassador:0.82.1-rc1 quay.io/datawire/ambassador:0.82.1-latest-rc | /home/lukeshu/src/github.com/datawire/ambassador/build-aux/bin/write-dockertagfile ambassador.docker.tag.release-rc

$ cat ambassador.docker.tag.release-rc
sha256:315bcf1703a1dbfbfc0476c6535b1c31e350c196516c9976a046eacb679a35c2
quay.io/datawire/ambassador:0.82.1-rc1
quay.io/datawire/ambassador:0.82.1-latest-rc

$ grep -A5 -B4 =GitInfo ambassador/ambassador/VERSION.py                                            
Version = "0.82.1"

Build = BuildInfo(
    version=Version,
    git=GitInfo(
        commit="ddc9f9e2",
        branch="lukeshu/versions",
        dirty=bool(""),
        description="v0.82.1-rc1",
    )
```

**Comment by kflynn** [comment]
Looks good from here...

**Pull request merged**